cpp/src/arrow/array.cc

```diff
@@ -48,13 +48,14 @@ Array::Array(const std::shared_ptr<DataType>& type, int32_t length,
       null_count_(null_count),
       null_bitmap_(null_bitmap),
       null_bitmap_data_(nullptr) {
+  if (null_count_ == 0) { null_bitmap_ = nullptr; }
```

Comment:
Acutally first thought this might be the issue (i.e. using a non-valid bitmap) but it wasn't. Still for `null_count == 0` we don't need it and leaving out the reference even makes the lazy branch of `null_count()` a lot faster.